addons/libs/dialog.lua

```diff
@@ -0,0 +1,158 @@
+-- This library was written to help find the ID of a known
+-- action message corresponding to an entry in the dialog tables.
+-- While the IDs can be collected in-game, they occasionally
+-- change and would otherwise need to be manually updated.
+-- It can also be used to find and decode an entry given the ID.
+
+-- Common parameters:
+-- dat: The entire content of the zone dialog DAT file
+-- i.e. local dat = io.open('path/to/dialog/DAT', 'rb'):read('*a')
+-- entry: The string you are looking for. If you do not know the
+-- entire string, use dev_find_substring.
+
+local xor = require('bit').bxor
+require('pack')
+local string = require('string')
+local unpack = string.unpack
+local pack = string.pack
+local find = string.find
+
+local dialog = {}
+
+local function decode(int)
+    return xor(int, 0x80808080)
+end
+local encode = decode
+
+local floor = require('math').floor
+local function binary_search(pos, dat, n)
+    local l, r, m = 1, n
+    while l < r do
+        m = floor((l+r)/2)
+        if decode(unpack('<I', dat, 1+4*m)) < pos then
+        -- offset given by mth ID < offset to string
+            l = m + 1
+        else
+            r = m
+        end
+    end
+    return l-2 -- we want the index to the left of where "pos" would be placed
+end
+
+local function plain_text_gmatch(text, substring, n)
+    n = n or 1
+    return function()
+        local pos = find(text, substring, n, true)
+        if pos then n = pos + 1 end
+        return pos
+    end
+end
+
+function dialog.dev_get_offset(dat, id) -- sanity check function
+    return string.format('%x', decode(unpack('<I', dat, 5+4*id)))
+end
+
+-- Returns the number of entries in the given dialog DAT file
+function dialog.entry_count(dat)
+    return decode(unpack('<I', dat, 5))/4
+end
+
+-- Returns an array-like table containing every ID which matched
+-- the given entry.
+-- An important note about this function: The tables contain an
+-- enormous number of duplicate entries. Additionally, some short
+-- entries can occur as substrings of longer entries. Results
+-- which fall into the second case should be filtered out.
+-- If you are dealing with the first case, it may help to look at
+-- any other messages which were received at the same time.
```

Comment:
Hm, not sure what you mean by that, can you give an example of a DAT where some entries are substrings of other entries? Or do you just mean that some string can occur in two different places, where one contains the other? And if so, what do you mean "should be filtered out"?

Reply:
The second: some entries were only a few bytes long and the same sequence of bytes showed up within others. The function matches by finding the given string within the file, verifying that the position is a valid offset, and that the length of the entry matches the length of the string. By being "filtered out" I meant that, if you were searching for a short entry, those longer results would not be added to the table returned by the function since the length of the entries would not match.

Reply:
Ok, I feel this is pretty self-evident. If I searched for "Hello" I wouldn't want it to return "Hello world".